### PR TITLE
Allow sorting blank forms by last saved

### LINKED ISSUE #5195: Allow sorting blank forms by last saved
As discussed [on the forum](https://forum.getodk.org/t/possibility-to-order-the-form-list-following-the-number-of-submission-most-used-on-top-in-collect/36524/6)

## Acceptance

- [ ] Given I have more than one blank form
And have recently filled one of them out (but not the others)
When I go to Fill Blank Form
And change sort order to "Last saved first"
Then the form I recently filled out moves to the top

- [ ] Given I have more than two blank forms
And I fill one out and then another
When I go to Fill Blank Form
And change sort order to "Last saved first"
Then the form I most recently filled out moves to the top
And the form I filled out before that is second in the list

### PR DESCRIPTION
**Pull request by grzesiek2010**
Closes #5195 

#### What has been done to verify that this works as intended?
I've tested the new option manually and added automated tests.

#### Why is this the best possible solution? Were any other approaches considered?
There is not much to discuss here. Theoretically we could add a new column to forms db to track last usage and then sort by that column but it would make the implementation more complicated and risky. That's why some time ago we refactored the list of forms to be able to implement sorting in the way I did here.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
This should just add a new sorting option. However I've implemented some code refactoring so please test all sorting options to make sure there is no regression.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Before submitting this PR, please make sure you have:
- [x] run `./gradlew checkAll` and confirmed all checks still pass OR confirm CircleCI build passes and run `./gradlew connectedDebugAndroidTest` locally.
- [x] verified that any code or assets from external sources are properly credited in comments and/or in the [about file](https://github.com/getodk/collect/blob/master/collect_app/src/main/assets/open_source_licenses.html).
- [x] verified that any new UI elements use theme colors. [UI Components Style guidelines](https://github.com/getodk/collect/blob/master/docs/CODE-GUIDELINES.md#ui-components-style-guidelines)

### CODE REVIEW

File: collect_app/src/main/java/org/odk/collect/android/activities/AppListActivity.java [hunk 1]
```diff
@@ -296,13 +296,10 @@ private void showBottomSheetDialog() {
...
-            @Override
-            public void onItemClicked(SortDialogAdapter.ViewHolder holder, int position) {
-                holder.updateItemColor(selectedSortingOrder);
-                performSelectedSearch(position);
-                bottomSheetDialog.dismiss();
-            }
+        final FormListSortingAdapter adapter = new FormListSortingAdapter(sortingOptions, getSelectedSortingOrder(), position -> {
+            performSelectedSearch(position);
+            bottomSheetDialog.dismiss();
+            return null;
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/activities/AppListActivity.java]: You can make `FormListSortingAdapter` take a `Consumer` (instead of a `(Int) -> Unit`) to prevent needing to do the `return null` here.

File: collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItem.kt [hunk 2]
```diff
@@ -12,3 +15,17 @@ data class BlankFormListItem(
     val dateOfLastUsage: Long,
     val contentUri: Uri
 )
+
+fun formToBlankFormListItem(form: Form, projectId: String, instancesRepository: InstancesRepository) = BlankFormListItem(
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItem.kt]: How would you feel about making this an extension function on `Form` instead?

File: collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    private val listener: (position: Int) -> Unit
+) : RecyclerView.Adapter<FormListSortingAdapter.ViewHolder>() {
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
+        val binding = SortItemLayoutBinding.inflate(LayoutInflater.from(parent.context), parent, false)
+        return ViewHolder(binding)
+    }
+
+    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
+        with(holder) {
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt]: I don't think you need this `with` as the contained code references `holder` directly anyway. Also, I find nesting functions like `with` makes it hard to track what is being referred to implicitly or with `this`.

File: collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+
+@RunWith(AndroidJUnit4::class)
+class BlankFormListItemTest {
+    private val instancesRepository = InMemInstancesRepository()
+
+    @Test
+    fun `Form should be properly converted to BlankFormListItem`() {
+        instancesRepository.save(
+            Instance.Builder()
+                .formId("0")
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt]: I found it a little hard to track the form ids here. I think it would be easier to read if the `Form` was created first and then referenced like:

```kotlin
instancesRepository.save(
            Instance.Builder()
                .formId(form.id)
                .formVersion(form.version)
...
```

Comment (seadowg) [reply collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt]: Same goes for the last test

**Review by seadowg — CHANGES_REQUESTED** [review]
I got disrupted in the middle of reviewing this, and I'll need to come back to it properly. Here are the comments I had so far though.

File: collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt (hunk 3 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt]: U used it to avoid repeating `holder.`
Do you like this version better?:
```
        holder.binding.root.setOnClickListener {
            listener.invoke(position)
            selectItem(holder.binding)
        }

        val sortingOption = sortingOptions[position]
        
        holder.binding.title.setText(sortingOption.text)
        holder.binding.icon.setImageResource(sortingOption.icon)
        holder.binding.icon.tag = sortingOption.icon
        holder.binding.icon.setImageDrawable(
            DrawableCompat.wrap(holder.binding.icon.drawable).mutate()
        )

        if (position == selectedSortingOrder) {
            selectItem(holder.binding)
        }
```

File: collect_app/src/main/java/org/odk/collect/android/activities/AppListActivity.java (hunk 1 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/activities/AppListActivity.java]: Right, fixed.

File: collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItem.kt (hunk 2 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItem.kt]: Yeah why not, done.

File: collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt (hunk 4 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt]: Ok, done.

File: collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt (hunk 3 above)
Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt]: I do actually find that easier to read. I think I find the double nesting hard to track and I do think I find `with` harder to follow that `apply`/`also`.

Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formlists/sorting/FormListSortingAdapter.kt]: I don't have any strong preferences so ok we can use that version.

File: collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt [hunk 5]
```diff
@@ -224,6 +207,7 @@ class BlankFormListViewModel(
             1 -> _allForms.value.sortedByDescending { it.formName.lowercase() }
             2 -> _allForms.value.sortedByDescending { it.dateOfCreation }
             3 -> _allForms.value.sortedBy { it.dateOfCreation }
+            4 -> _allForms.value.sortedByDescending { it.dateOfLastUsage }
```
Comment (seadowg) [inline collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt:210]: I think a really nice change we could make while we're in here would be to remove any knowledge of the display order of these sort options from `BlankFormListViewModel`. One thing that comes to mind is that we could have an `enum class` of sort "types" and an additional property on `FormListSortingOption` of that. Then, `FormListSortingAdapter`'s listener could receive that enum instead of the index which would allow this `when` to match on the sort type rather than the index.

I think a change like the above would make the code and tests easier to read as it wouldn't require any context of the display order.

File: collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+import org.odk.collect.forms.instances.Instance
+import org.odk.collect.formstest.InMemInstancesRepository
+import org.odk.collect.projects.Project
+
+@RunWith(AndroidJUnit4::class)
+class BlankFormListItemTest {
+    private val instancesRepository = InMemInstancesRepository()
+
+    @Test
+    fun `Form should be properly converted to BlankFormListItem`() {
```
Comment (seadowg) [inline collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt:19]: It looks like you do enough work in the other tests in this file to just have a single form and instance here to check a `Form` to `BlankFormListItem` conversion where everything is populated. I might have missed an edge case this is capturing, but if so I think that should have its own test.

**Review by seadowg — CHANGES_REQUESTED** [review]

File: collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt (hunk 6 above)
Comment (grzesiek2010) [reply collect_app/src/test/java/org/odk/collect/android/formlists/blankformlist/BlankFormListItemTest.kt:19]: Right after adding more tests in this class this one seems to be to big. Fixed.

File: collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt (hunk 5 above)
Comment (grzesiek2010) [reply collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt:210]: Ohh this would require a lot of changes and I'm not sure if it is worth it...
`FormListSortingAdapter` is supposed to be a class for all types of lists and different types of lists have different sorting options I think representing them by enums would require mixing all of them in one class.

I would need to have more time to warm up to this idea but I'm off tomorrow and on Friday so won't have time to do that. If you are sure this would make the code better please take over.

Comment (seadowg) [reply collect_app/src/main/java/org/odk/collect/android/formlists/blankformlist/BlankFormListViewModel.kt:210]: Good points. Sounds like something we can come back to if we're in here again. No need to rush to block this at the moment.

**Review by seadowg — APPROVED** [review]

**Comment by dbemke** [comment]
(sort by last saved first) When hide old form versions in disabled in settings and a user adds a new version of a form or adds a form on Central in Form access, the forms are at the bottom of the list of forms (after alphabetically ordered forms which were on the list before) in Fill Blank Form.
If afterwards the user enables hide old form version the new version of forms go back to being sorted alphabetically, but the form added to list on Central stays at the bottom of the list.

**Steps to reproduce:**

1. In Collect settings disable hide old form versions.
2. Go to Fill Blank Form.
3. Choose sort by Last saved first.
4. Fill in a form.
5. Go to Central and add new version of the form.
6. Refresh the list of forms in Fill Blank Form

**Steps to reproduce:**

1. In Collect settings disable hide old form versions.
2. Go to Fill Blank Form.
3. Choose sort by Last saved first.
4. Add access to a form on Central in Form access.
5. Refresh the list of forms in Fill Blank Form

**Comment by dbemke** [comment]
I’m not really sure what should be the expected result when sort by last saved is chosen and a user deletes the saved forms.  There is a difference between finalized and not finalized form. When a finalized form is deleted it stays at the top of the list in Fill Blank Form. When a non-finalized form is deleted it goes back to the (probably) alphabetic order – is no longer at the top of the list in Fill Blank Form.

**Steps to reproduce:**

1. In Collect settings set blank form update mode to manual.
2. Go to Fill Blank Form.
3. Choose sort by Last saved first.
4. Fill in one form and finalize it.
5. Fill in a different form but don’t finalize it, just save and exit.
6. Go to Delete Saved Form.
7. In Saved Forms select the two forms and delete them.
8. Go back to Fill Blank Form.

**Comment by grzesiek2010** [comment]
> I’m not really sure what should be the expected result when sort by last saved is chosen and a user deletes the saved forms. There is a difference between finalized and not finalized form. When a finalized form is deleted it stays at the top of the list in Fill Blank Form. When a non-finalized form is deleted it goes back to the (probably) alphabetic order – is no longer at the top of the list in Fill Blank Form.

It's not a difference between `not finalized` vs `finalized` but between `unsent` vs `sent` (you probably have `autosend` enabled so your finalized forms are sent at the same time). If a form is unsent and we remove it in the way you described it is removed entirely from the `instances` database which is used to determine the order of last used forms. If a form is sent even if you remove it it is still in our database because we need it for example to preview in `View Sent Form`.
I think this behavior is ok.

**Comment by grzesiek2010** [comment]
> (sort by last saved first) When hide old form versions in disabled in settings and a user adds a new version of a form or adds a form on Central in Form access, the forms are at the bottom of the list of forms (after alphabetically ordered forms which were on the list before) in Fill Blank Form.
If afterwards the user enables hide old form version the new version of forms go back to being sorted alphabetically, but the form added to list on Central stays at the bottom of the list.

Are there any filled forms in those scenarios? Or just blank forms without any instances as described in the steps?

**Comment by dbemke** [comment]
In "Delete Saved Form" a chosen option in "sort by" stays highlighted even if the user taps on a different option- a few sort by options can be highlighted at the same time. Although a few sort by options are highlighted, the chosen option (e.g A-Z) sorts the list as expected. The issue is not present on master version.
sortdeletesaved

**Steps to reproduce:**
1. Fill and save some forms.
2. Go to Delete Saved Form.
3. Tap on sort by button.
4. Choose one of the options.
5. Tap again on a sort by button.
6. Choose a different option.

**Comment by grzesiek2010** [comment]
@dbemke the issue should be fixed.

**Comment by dbemke** [comment]
**Tested with Success!**

Verified on Android 10, 13 (physical device and the last commit on the emulator)

**Verified Cases:**
- issue with highlighted sort by option is no longer reproducing
- saving via save button, save and exit button, device back button+ save changes
- deleting a saved (finalized, not finalized) form
- switching between different sort by options
- saving a form entered through form map view, a shortcut
- sort by last saved with new versions of a form with hide old versions enabled and disabled
- sort by last saved with new forms added to a project
- saveIncomplete form + ignoring changes
- switching between different sort by options in Fill Blank Form, Edited Saved Form, Send Finalized Form, View Sent Form, Get Blank Form, Delete Saved Form
